components/keys/src/types.rs

```diff
@@ -34,25 +36,33 @@ impl Key {
     /// Creates a key from raw bytes.
     #[inline]
     pub fn from_raw(key: &[u8]) -> Key {
-        // adding extra length for appending timestamp
-        let len = codec::bytes::max_encoded_bytes_size(key.len()) + codec::number::U64_SIZE;
+        let len = MemComparableByteCodec::encoded_len(key.len());
         let mut encoded = Vec::with_capacity(len);
-        encoded.encode_bytes(key, false).unwrap();
+        encoded.write_comparable_bytes(key).unwrap();
         Key(encoded)
     }
 
     /// Gets and moves the raw representation of this key.
     #[inline]
     pub fn into_raw(self) -> Result<Vec<u8>, codec::Error> {
         let mut k = self.0;
-        bytes::decode_bytes_in_place(&mut k, false)?;
+        MemComparableByteCodec::try_decode_first_in_place(&mut k)?;
+        k.shrink_to_fit();
```

Comment:
I don't think shrink_to_fit will work. It only shrinks extra capacity while in this case there is extra data in length.

Reply:
`shrink_to(written_bytes)`?

Reply:
you may want the truncate method.